src/server/master_server/metadata_manager.cc

```diff
@@ -0,0 +1,102 @@
+#include "metadata_manager.h"
+#include <stack>
+
+namespace gfs {
+namespace server {
+
+MetadataManager::MetadataManager() {
+  lockManager_ = LockManager::GetInstance();
+}
+
+bool MetadataManager::CreateFileMetadata(const std::string& pathname) {
+  // Step 1. Lock (readerlock underneath) the parent directory first
+  ParentLocksAnchor parentLockAnchor(lockManager_, pathname);
+             
+  if (!parentLockAnchor.succ()) {
+    // If this operation fails, which means some of the parent directory does not
+    // exist, we return false
+    return false;
+  }
+
+  // Step 2. Add a new lock for this new file, and writeLock it
+  auto newLock(lockManager_->AddLockIfNonExist(pathname));
+  if (!newLock) {
+    // We still need to check whether the return is NULL because another thread
+    // could well successfully created a new lock for the same path
+    return false;
+  }
+   
+  absl::WriterMutexLock anchorForNewLock(newLock);
+  // Step 3. writeLock the global lock, instantiate a FileMetadata object
+  absl::WriterMutexLock anchorForGlobalLock(lockManager_->globalLock());
+  // The reason that we acquire the global lock is that we need to synchronization
+  // between write and read from the fileMetadata collection.
+  if(fileMetadata_.count(pathname)) {
+    return false;
+  }
+   
+  fileMetadata_[pathname] = std::make_shared<FileMetadata>();
+  return true;
+}
+
+bool MetadataManager::ExistFileMetadata(const std::string& pathname) const {
+  absl::ReaderMutexLock(lockManager_->globalLock());
+  return fileMetadata_.count(pathname);
+}
+
+std::shared_ptr<FileMetadata> MetadataManager::GetFileMetadata(
+  const std::string& pathname) const {
+  // readLock the global lock and retrieve the filemetadata. The reason for a
+  // readLock is because we are not mutating anything in the fileMetadata_.
+  absl::ReaderMutexLock(lockManager_->globalLock());
+  return fileMetadata_.at(pathname);
```

Comment:
What happened if we don't have this `pathname` in fileMetadata?

Reply:
The same issue gets addresed by StatusOr